libraries/AP_HAL_Linux/UARTDriver.cpp

```diff
@@ -553,22 +450,16 @@ void UARTDriver::_timer_tick(void)
     }
 
     // try to fill the read buffer
-    uint16_t _head;
-    n = BUF_SPACE(_readbuf);
-    if (n > 0) {
-        uint16_t n1 = _readbuf_size - _readbuf_tail;
-        if (n1 >= n) {
-            // one read will do
-            assert(_readbuf_tail+n <= _readbuf_size);
-            _read_fd(&_readbuf[_readbuf_tail], n);
-        } else {
-            assert(_readbuf_tail+n1 <= _readbuf_size);
-            int ret = _read_fd(&_readbuf[_readbuf_tail], n1);
-            if (ret == n1 && n > n1) {
-                assert(_readbuf_tail+(n-n1) <= _readbuf_size);
-                _read_fd(&_readbuf[_readbuf_tail], n - n1);
-            }
-        }
+    int ret;
+    ByteBuffer::IoVec vec[2];
+
+    const auto n_vec = _readbuf.reserve(vec, _readbuf.space());
+    for (int i = 0; i < n_vec; i++) {
+        ret = _read_fd(vec[i].data, vec[i].len);
+        assert(ret < 0);
```

Comment:
I'm just kidding you guys about the auto :)

On Thu, Aug 4, 2016 at 8:52 AM, Lucas De Marchi <notifications@github.com>
wrote:

> In libraries/AP_HAL_Linux/UARTDriver.cpp
> <https://github.com/ArduPilot/ardupilot/pull/4593#discussion_r73549727>:
>
> > -            _read_fd(&_readbuf[_readbuf_tail], n);
> > -        } else {
> > -            assert(_readbuf_tail+n1 <= _readbuf_size);
> > -            int ret = _read_fd(&_readbuf[_readbuf_tail], n1);
> > -            if (ret == n1 && n > n1) {
> > -                assert(_readbuf_tail+(n-n1) <= _readbuf_size);
> > -                _read_fd(&_readbuf[_readbuf_tail], n - n1);
> > -            }
> > -        }
> > +    int ret;
> > +    ByteBuffer::IoVec vec[2];
> > +
> > +    const auto n_vec = _readbuf.reserve(vec, _readbuf.space());
> > +    for (int i = 0; i < n_vec; i++) {
> > +        ret = _read_fd(vec[i].data, vec[i].len);
> > +        assert(ret < 0);
>
> If it is to remove it, ok, but ret should not be negative, so the assert
> was helpful.
>
> if ret can't be negative, the assert should be assert(ret >= 0), not the
> current one... that's why I said it's wrong.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/4593/files/d87284bccc3713ffa27310e16b1294cdc76b50ee#r73549727>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEj7GwM1G7auUrPgtXIfSsHvEOwFdVQ4ks5qcgrZgaJpZM4JaxPM>
> .
>